### ISSUE TITLE
On-device sampling corrupts all but one sequence slot at max_num_seqs > 1 on Inf2 (silent, HTTP 200)

### ISSUE THREAD

**Issue by AbrahamArellano** [vllm-project/vllm-neuron#57]
## Summary

With on-device sampling enabled (the default), `max_num_seqs > 1` on the contiguous KV path returns **silently corrupted output for all but one sequence slot**. Every request returns HTTP 200 with a well-formed OpenAI response; the text is wordlike gibberish rather than an error, so nothing downstream can detect it.

Disabling on-device sampling fixes it completely. That workaround is currently unreachable because of the one-line import defect in #29, which is still present on `main`.

**This is a correctness bug, not a security issue.** No trust boundary is crossed, no attacker input is required or helpful, and no data leaks between requests: each slot returns garbage derived from its own prompt. Filing publicly per CONTRIBUTING rather than through the vulnerability reporting path.

## Environment

- Instance: `inf2.xlarge` (1 Neuron device, 2 cores), TP 2
- Model: `Qwen/Qwen3-1.7B` @ `70d244cc86ccca08cf5af4e1e306ecf908b1ad5e`
- Images (AWS DLC `public.ecr.aws/neuron/pytorch-inference-vllm-neuronx`):
  - `0.13.0-neuronx-py312-sdk2.28.0-ubuntu24.04` → `vllm-neuron` **0.4.1**, vLLM 0.13.0, NxDI `0.8.16251`
  - `0.16.0-neuronx-py312-sdk2.31.0-ubuntu24.04` → `vllm-neuron` **0.5.3**, vLLM 0.16.0, NxDI `0.10.18399`
- Flags: `--max-model-len 2048 --no-enable-prefix-caching --num-gpu-blocks-override <max_num_seqs>` (equal to `max_num_seqs`, as the NxD Inference vLLM guide requires on the contiguous KV path)

## Reproduction

Serve with `--max-num-seqs 4`, then send the **same** prompt 8 times serially with greedy decoding (`temperature: 0.0`), and 4 times concurrently.

```bash
vllm serve /models/qwen3-1.7b \
  --tensor-parallel-size 2 --max-model-len 2048 --max-num-seqs 4 \
  --no-enable-prefix-caching --num-gpu-blocks-override 4 --port 8000
```

## Results

Legibility judged by content (non-ASCII share, top 3-gram repetition consulted only at >= 30 3-grams, alphabetic-token share; any one condemns).

| plugin | model source | `max_num_seqs` | serial legible | concurrent legible |
|---|---|---|---|---|
| 0.4.1 | local checkpoint | 4 | 2/8 | 1/4 |
| 0.5.3 | local checkpoint | 4 | 3/8 | 1/4 |
| 0.5.3 | HF model id | 4 | 3/8 | 1/4 |
| 0.5.3 | local checkpoint | **1** | **8/8** | **4/4** |
| 0.5.3 + import fix + sampling disabled | local checkpoint | 4 | **8/8** | **4/4** |

Representative corrupted output (same prompt that yields a correct answer in the working slot):

```
Certainly(+okokok.pok.p.p.p.piesok.p.pies.pies.piesiesieson
Certainly nausea nausea,rint,il* nausea*ist*ist(/l(*...
```

### The signature is periodic in the slot count

On 0.4.1 the legible answers are requests 1 and 5, exactly four apart, and the corrupted answers pair off by the same period: request 2 is byte-identical to 6, 3 to 7, and 4 to 8. That matches the four sequence slots of the contiguous KV layout, with only one slot returning correct tokens. On 0.5.3 the signature is perturbed but not fixed (a second slot recovers).

## Analysis

`get_neuron_model` builds on-device sampling unconditionally:

```python
on_device_sampling_config = OnDeviceSamplingConfig(dynamic=True, deterministic=False)
```

NxD Inference then samples inside the compiled NEFF and returns tokens rather than logits. Across a multi-slot decode batch that sampling appears misaligned with the slots, so one slot yields correct tokens and the rest yield gibberish.

Three observations support this and rule out the alternatives we tested:

1. `max_num_seqs 1` is always correct, so with a single slot there is nothing to misalign.
2. Model source is irrelevant (local checkpoint and HF id give identical results), so the documented `tie_word_embeddings` local-checkpoint remedy does not address this.
3. With on-device sampling disabled, all eight answers are **byte-identical**, which is what greedy decoding should produce and what `deterministic=False` on-device sampling would not.

We did **not** establish the mechanism inside NxDI, since the sampling runs in the compiled NEFF. What is shown is that framework sampling avoids the fault, not why on-device sampling misaligns.

## Workaround, and why it is currently blocked

All three parts are required; any one alone does nothing:

1. The one-line import fix from #29 / #28 (still wrong on `main`):
   ```python
   from vllm.v1.sample.sampler import Sampler   # not: from vllm.v1.sample import sampler as Sampler
   ```
2. `NEURON_ON_DEVICE_SAMPLING_DISABLED=1`
3. `--additional-config '{"override_neuron_config": {"on_device_sampling_config": null}}'`

Part 2 alone is insufficient and we measured that case: it only constructs a framework sampler while `get_neuron_model` still builds the NxDI config, and the result was unchanged at 3/8 and 1/4.

Packaging note for anyone applying the patch to a DLC image: `vllm_neuron` ships three times (`/opt/conda/lib/python3.12/site-packages`, `/opt/vllm`, `/opt/vllm/build/lib`) and the active import is the site-packages copy.

## Ask

1. Reopen and merge the one-line import fix (#29, PR #28), on the basis that it gates a correctness workaround rather than a convenience flag.
2. Treat the slot misalignment under on-device sampling as a correctness bug on the contiguous KV path, or document `max_num_seqs 1` as the only supported setting there.
3. If neither is planned for the Inf2-capable line, saying so plainly would let operators choose framework sampling deliberately instead of discovering silent corruption in production.

## Scope

Measured on **Inf2 only**, on plugin lines **0.4.1** and **0.5.3**. We have no Trn2 or Trn3 hardware and make **no claim** about the 0.21 or 0.24 lines. The import defect specifically was verified as still present on `main`.
